python/cudf/udf_cpp/nrt/_nrt_cuda.cuh

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright (c) 2012, Anaconda, Inc.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are
+met:
+
+Redistributions of source code must retain the above copyright notice,
+this list of conditions and the following disclaimer.
+
+Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <cuda/atomic>
+
+typedef __device__ void (*NRT_dtor_function)(void* ptr, size_t size, void* info);
+typedef __device__ void (*NRT_dealloc_func)(void* ptr, void* dealloc_info);
+
+typedef __device__ void* (*NRT_malloc_func)(size_t size);
+typedef __device__ void* (*NRT_realloc_func)(void* ptr, size_t new_size);
+typedef __device__ void (*NRT_free_func)(void* ptr);
+
+extern "C" {
+struct MemInfo {
+  cuda::atomic<size_t, cuda::thread_scope_device> refct;
+  NRT_dtor_function dtor;
+  void* dtor_info;
+  void* data;
+  size_t size;
+};
+}
+
+typedef struct MemInfo NRT_MemInfo;
+
+// Globally needed variables
+struct NRT_MemSys {
+  /* System allocation functions */
+  struct {
+    NRT_malloc_func malloc;
+    NRT_realloc_func realloc;
+    NRT_free_func free;
+  } allocator;
+  struct {
+    bool enabled;
+    cuda::atomic<size_t, cuda::thread_scope_device> alloc;
+    cuda::atomic<size_t, cuda::thread_scope_device> free;
+    cuda::atomic<size_t, cuda::thread_scope_device> mi_alloc;
+    cuda::atomic<size_t, cuda::thread_scope_device> mi_free;
+} stats;
+};
+
+/* The Memory System object */
+__device__ NRT_MemSys TheMSys;
+
+extern "C" __device__ void* NRT_Allocate(size_t size) { return malloc(size); }
+
+extern "C" __device__ void NRT_MemInfo_init(
+  NRT_MemInfo* mi, void* data, size_t size, NRT_dtor_function dtor, void* dtor_info)
+{
+  mi->refct     = 1; /* starts with 1 refct */
+  mi->dtor      = dtor;
+  mi->dtor_info = dtor_info;
+  mi->data      = data;
+  mi->size      = size;
+  if (TheMSys.stats.enabled)
+  {
+    TheMSys.stats.mi_alloc++;
+  }
+}
+
+__device__ NRT_MemInfo* NRT_MemInfo_new(void* data,
```

Comment:
This is vendored from [nrt.cpp](https://github.com/numba/numba/blob/main/numba/core/runtime/nrt.cpp#L183) where debug statements have been removed and the call to `NRT_MemInfo_init` has been adjusted to align with its new signature that drops the external allocator.